### PR TITLE
chore: adds storage tests for upgrade

### PR DESCRIPTION
**Pull request by cridmann**
Adds `SuperstateTokenStorageLayoutTestBase` which is flexible enough to support all of our future token upgrades with maximal code reuse.

Within `USTBv2TokenStorageLayoutTests` we are calling out a known issue w/ the upgrade from `v1` to `v2` wrt storage of `nonces`, `encumberedBalanceOf`, and `encumbrances`. This is accepted because these features have not been used in our deployed contracts yet. They will work as expected after upgrading.

Also, this PR adds protections for upgrades in the future by adding the following to `v2` of `SuperstateToken`:
- `uint256[500] private __inheritanceGap` - added as the first storage field within `SuperstateToken`, which will allow us to inherit more contracts in the future. This value is quite large to support contracts that have their own large storage gap (which is usually around 50 slots).
-`uint256[100] private __additionalFieldsGap;` - added as the last storage field within `SuperstateToken`, to allow us to add other fields in future upgrades without impacting other contracts that inherit from `SuperstateToken`.

### CODE REVIEW

File: test/SuperstateTokenStorageLayoutTestBase.t.sol [hunk 1]
```diff
@@ -0,0 +1,292 @@
...
+
+    string public oldTokenVersion;
+    string public newTokenVersion;
+
+    address public alice = address(10);
+    address public bob = address(11);
+    address public charlie = address(12);
+    address public mallory = address(13);
+    uint256 evePrivateKey = 0x353;
+    address eve; // see setup()
```
Comment (jonwalch) [inline test/SuperstateTokenStorageLayoutTestBase.t.sol:38]:
```suggestion
    address public eve; // see setup()
```

File: test/v2/USTBv2TokenStorageLayoutTests.t.sol [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+import {USTBv1} from "src/v1/USTBv1.sol";
+import {USTB} from "src/USTB.sol";
+import {AllowList} from "src/AllowList.sol";
+import "test/AllowListV2.sol";
+import "test/USTBV2.sol";
+import "test/SuperstateTokenStorageLayoutTestBase.t.sol";
+import {console} from "forge-std/console.sol";
+
+/**
+ *  SuperstateV1 Token storage layout:
```
Comment (cridmann) [inline test/v2/USTBv2TokenStorageLayoutTests.t.sol:22]: shows v1 storage slots

File: test/v2/USTBv2TokenStorageLayoutTests.t.sol [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+ *  Slot 54: Erc20Upgradeable._name
+ *  Slot 55: Erc20Upgradeable._symbol
+ *  Slot 101: PausableUpgradeable._paused
+ *  Slot 151: SuperstateToken.nonces
+ *  Slot 152: SuperstateToken.encumberedBalanceOf
+ *  Slot 153: SuperstateToken.encumbrances
+ *  Slot 154: SuperstateToken.accountingPaused
+ *
+ *
+ *  SuperstateV2 Token storage layout:
```
Comment (cridmann) [inline test/v2/USTBv2TokenStorageLayoutTests.t.sol:36]: shows v2 storage slots

File: test/v2/USTBv2TokenStorageLayoutTests.t.sol [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+            assertEq(currentToken.encumbrances(alice, bob), expectedAliceEncumberedBalanceOf);
+
+            // assert accountingPaused (storage slot 154)
+            uint256 accountingPausedSlotValue = uint256(loadSlot(154));
+            assertEq(1, accountingPausedSlotValue);
+
+            // assert accountingPaused from contract method
+            assertEq(true, currentToken.accountingPaused());
+
+            // assert decimals
```
Comment (jonwalch) [inline test/v2/USTBv2TokenStorageLayoutTests.t.sol:159]: RM or TODO?

**Comment by jonwalch** [comment]
"Within USTBv2TokenStorageLayoutTests we are calling out a known issue w/ the upgrade from v1 to v2 wrt storage of nonces, encumberedBalanceOf, and encumbrances. This is accepted because these features have not been used in our deployed contracts yet."

will work as expected after upgrading?

**Review by maxwolff — APPROVED** [review]

File: test/v2/USTBv2TokenStorageLayoutTests.t.sol (hunk 4 above)
Comment (cridmann) [reply test/v2/USTBv2TokenStorageLayoutTests.t.sol:159]: Replaced with `// ignore decimals due to being in-lined in bytecode`

**Comment by cridmann** [comment]
`will work as expected after upgrading?` yes

**Pull request merged**